contrib/DC-GAN.ipynb

```diff
@@ -0,0 +1,525 @@
+    "\n",
+    "<br> At every training step, each of the cores perfoms the forward (loss computation) and backward (gradient computation) on the given minibatch and then [all_reduce](https://www.tensorflow.org/xla/operation_semantics#allreduce) is performed across TPU cores to update the parameters. Notice `xm.optimizer_step` call in the discriminator and optimizer train steps.\n",
+    "\n",
+    "\n",
+    "General GAN training looks like:\n",
+    "\n",
+    "1. update the **generator** ($G$) to minimize the probability of the __discriminator making the correct choice__. \n",
+    "2. update the **discriminator** ($D$) to maximize the probability of the __discriminator making the correct choice__.\n",
+    "\n",
+    "We will use a different objective when we update the generator: maximize the probability of the **discriminator making the incorrect choice**. This small change helps to allevaiate problems with the generator gradient vanishing when the discriminator is confident. This is the standard update used in most GAN papers, and was used in the original paper from [Goodfellow et al.](https://arxiv.org/abs/1406.2661). \n",
+    "\n",
+    "Therefore the training lopp in this notebook will entail:\n",
+    "1. Update the generator ($G$) to maximize the probability of the discriminator making the incorrect choice on generated data:\n",
+    "$$\\underset{G}{\\text{maximize}}\\;  \\mathbb{E}_{z \\sim p(z)}\\left[\\log D(G(z))\\right]$$\n",
+    "2. Update the discriminator ($D$), to maximize the probability of the discriminator making the correct choice on real and generated data:\n",
+    "$$\\underset{D}{\\text{maximize}}\\; \\mathbb{E}_{x \\sim p_\\text{data}}\\left[\\log D(x)\\right] + \\mathbb{E}_{z \\sim p(z)}\\left[\\log \\left(1-D(G(z))\\right)\\right]$$\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import torch\n",
+    "from torch import nn, optim\n",
+    "from torch.autograd.variable import Variable\n",
+    "from torchvision import transforms, datasets\n",
+    "from torch.optim import Adam\n",
+    "import torch.nn.functional as F\n",
+    "\n",
+    "import torch_xla\n",
+    "import torch_xla.core.xla_model as xm\n",
+    "import torch_xla.debug.metrics as met\n",
+    "import torch_xla.distributed.parallel_loader as pl\n",
+    "import torch_xla.distributed.xla_multiprocessing as xmp\n",
+    "import torch_xla.utils.utils as xu\n",
+    "\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Setting up the Global Flags\n",
+    "\n",
+    "In the current setup, Discriminator network was chosen to be a smaller capacity than generator. Even with similar capacity netowrks, generator update path is deeper than Discriminator. Therefore uneven learning rates chosen here seems to yield a better convergence. \n"
```

Comment:
typo: netowrk - > network

Reply:
generator/Discriminator, one is small case, the other capital case, looks weird.

Reply:
corrected.